README.md
<p align="center">
  <img src="https://github.com/ipfs-shipyard/npm-on-ipfs/raw/master/img/npm-on-ipfs.jpg" alt="npm distributed on top of lots of connected IPFS nodes worldwide" />
</p>

# npm-on-IPFS

[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](https://protocol.ai)
[![](https://img.shields.io/badge/project-IPFS-blue.svg?style=flat-square)](http://ipfs.io/)
[![](https://img.shields.io/badge/freenode-%23ipfs-blue.svg?style=flat-square)](http://webchat.freenode.net/?channels=%23ipfs)
[![Build Status](https://flat.badgen.net/travis/ipfs-shipyard/npm-on-ipfs)](https://travis-ci.com/ipfs-shipyard/npm-on-ipfs)
[![Code Coverage](https://codecov.io/gh/ipfs-shipyard/npm-on-ipfs/branch/master/graph/badge.svg)](https://codecov.io/gh/ipfs-shipyard/npm-on-ipfs)
[![Dependency Status](https://david-dm.org/ipfs-shipyard/npm-on-ipfs.svg?style=flat-square)](https://david-dm.org/ipfs-shipyard/npm-on-ipfs)

**TLDR: npm-on-ipfs enables you to install your favourite modules from the distributed web using IPFS, as well as to have a cache always ready and shared on your local network — great for enterprise and community coding settings, or even just enabling more speedy work when you and your friends are working at a low-bandwidth coffee shop.**

## Quick background

As the largest software registry in the world, [npm](https://www.npmjs.com) is also the de facto package manager for the JavaScript ecosystem, with more than [900k](https://replicate.npmjs.com/_all_docs) packages and more than 7 billion downloads a week. It's incredibly fast and reliable — however, we couldn't stop ourselves from wondering what would happen if we put the world's largest registry on the distributed web.

The result is npm-on-ipfs: a module that wraps your package manager of choice (npm or yarn) in configuration to use [IPFS](https://ipfs.io/), not HTTP, to retrieve your dependencies from the central npm registry. It's still a work in progress, but we think you'll find it useful and awesome for the following reasons:

 - Having dependencies on the distributed web makes development **more available** because multiple nodes supplying tarballs means no panic if a single source goes dark
 - It can also be **faster and cheaper** — if dependencies are already being hosted on your local network, this means lower bandwidth cost and higher speed 
 - If enough dependencies are hosted on your local network (think enterprise or community development settings), that network can operate **offline-first**: Take your team on a remote mountain retreat and hack away!

## Install & use

```console
$ npm i ipfs-npm -g
```

### Get started!

`ipfs-npm` wraps your favorite package manager (npm or yarn) with configuration that uses IPFS, rather than HTTP, to retrieve your dependencies from the central npm registry. Since it's intended to replace npm/yarn, all the commands you're used to will work in the same way.

For example: In the directory with your `package.json` file, run ...

```console
$ ipfs-npm install
👿 Spawning an in-process IPFS node
Swarm listening on /ip4/127.0.0.1/tcp/57029/ipfs/QmVDtTRCoYyYu5JFdtrtBMS4ekPn8f9NndymoHdWuuJ7N2
🗂️ Loading registry index from https://registry.js.ipfs.io
☎️ Dialling registry mirror /ip4/35.178.192.119/tcp/10015/ipfs/QmWBaYSnmgZi6F6D69JuZGhyL8rm6pt8GX5r7Atc6Gd7vR,/dns4/registry.js.ipfs.io/tcp/10015/ipfs/QmWBaYSnmgZi6F6D69JuZGhyL8rm6pt8GX5r7Atc6Gd7vR
🗑️ Replacing old registry index if it exists
📠 Copying registry index /ipfs/QmQmVsNFw3stJky7agrETeB9kZqkcvLSLRnFFMrhiR8zG1 to /npm-registry
👩‍🚀 Starting local proxy
🚀 Server running on port 57314
🎁 Installing dependencies with /Users/alex/.nvm/versions/node/v10.8.0/bin/npm
...
```

You can use any command you'd use with npm/yarn with ipfs-npm in exactly the same way:

```console
$ ipfs-npm install
$ ipfs-npm version minor
$ ipfs-npm publish
$ ipfs-npm --package-manager=yarn
// etc
```

## CLI guide

```console
$ ipfs-npm --help
ipfs-npm

Installs your js dependencies using IPFS

Options:
  --help                        Show help                              [boolean]
  --version                     Show version number                    [boolean]
  --package-manager             Which package manager to use - eg. npm or yarn
                                                                [default: "npm"]
  --ipfs-registry               Where to download any packages that haven't made
                                it into the registry index yet from
                                        [default: "https://registry.js.ipfs.io"]
  --registry-upload-size-limit  How large a file upload to allow when proxying
                                for the registry             [default: "1024MB"]
  --registry-update-interval    Only request the manifest for a given module
                                every so many ms                [default: 60000]
  --registry-connect-timeout    How long to wait while dialling the mirror
                                before timing out                [default: 5000]
  --registry-read-timeout       How long to wait for individual packages before
                                timing out                       [default: 5000]
  --ipfs-mfs-prefix             Which mfs prefix to use
                                                      [default: "/npm-registry"]
  --ipfs-node                   "proc" to start an in-process IPFS node,
                                "disposable" to start an in-process disposable
                                node, "go" or "js" to spawn an IPFS node as a
                                separate process or a multiaddr that resolves to
                                a running node                 [default: "proc"]
  --ipfs-repo                   If --ipfs-node is set to "proc", this is the
                                path that contains the IPFS repo to use
                                                [default: "/Users/alex/.jsipfs"]
  --ipfs-flush                  Whether to flush the MFS cache   [default: true]
  --clone-pin                   Whether to pin cloned modules   [default: false]
  --request-max-sockets         How many concurrent http requests to make while
                                cloning the repo                   [default: 10]
  --request-retries             How many times to retry when downloading
                                manifests and tarballs from the registry
                                                                    [default: 5]
  --request-retry-delay         How long in ms to wait between retries
                                                                 [default: 1000]
  --request-timeout             How long in ms we should wait when requesting
                                files                           [default: 30000]
  --npm-registry                A fallback to use if the IPFS npm registry is
                                unavailable
                                         [default: "https://registry.npmjs.com"]
```

## Configuration files

ipfs-npm uses [`rc`](https://github.com/dominictarr/rc) to parse configuration files. Please see the [`rc` repository](https://github.com/dominictarr/rc#standards) for the order of precedence used when searching for configuration files. The app is `ipfs-npm`.

For instance, if you want to always use a remote daemon, you could create a `~/.ipfs-npmrc` file like this:

```json
{
  "ipfsNode": "/ip4/127.0.0.1/tcp/5001"
}
```

## To learn more

[Protocol Labs](https://protocol.ai), the organization behind IPFS, is actively working on improving the landscape for package managers and the distributed web in 2019 and beyond. To that end, we've created an [IPFS Package Managers Special Interest Group](https://github.com/ipfs/package-managers), and your feedback and contributions are very welcome!

If you're actively (or just casually) using npm-on-ipfs and have feedback about your user experience, we'd love to hear from you, too. Please open an issue in the [Special Interest Group](https://github.com/ipfs/package-managers) and we'll get right back to you.

More resources you may find useful:
- [The original npm-on-ipfs demo video](https://vimeo.com/147968322)
- [A more detailed introduction to npm-on-ipfs from David Dias' blog](http://daviddias.me/blog/stellar-module-management/)
- [Node.js Interactive talk on Stellar Module Management, aka npm-on-ipfs](https://www.youtube.com/watch?v=-S-Tc7Gl8FM)

## Lead maintainer

[Alex Potsides](https://github.com/achingbrain)


CHANGELOG.md
<a name="0.16.3"></a>
## [0.16.3](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.16.2...v0.16.3) (2019-06-06)


### Bug Fixes

* do not stop external IPFS node ([#99](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/99)) ([40c12a6](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/40c12a6))
* dont spawn node for local ops ([#110](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/110)) ([36f4e8f](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/36f4e8f))
* only spawn node once ([#111](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/111)) ([8741a66](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/8741a66))
* remove redundant ipfs configing code ([55978ab](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/55978ab))
* tests by adding go-ipfs to PATH ([#100](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/100)) ([7489c94](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/7489c94))


### Features

* support config files ([#102](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/102)) ([877840b](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/877840b))



<a name="0.16.2"></a>
## [0.16.2](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.16.1...v0.16.2) (2019-01-18)



<a name="0.16.1"></a>
## [0.16.1](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.16.0...v0.16.1) (2018-12-10)


### Bug Fixes

* catch errors when connecting to the registry ([a412f98](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/a412f98))



<a name="0.16.0"></a>
# [0.16.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.15.0...v0.16.0) (2018-12-07)


### Features

* inherit stdin so we can interact with package manager ([267b73d](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/267b73d))



<a name="0.15.0"></a>
# [0.15.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.14.0...v0.15.0) (2018-12-07)



<a name="0.14.0"></a>
# [0.14.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.13.0...v0.14.0) (2018-12-07)



<a name="0.13.0"></a>
# [0.13.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.12.0...v0.13.0) (2018-12-07)


### Features

* default to yarn for yarn and npm for npm ([011baad](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/011baad))



<a name="0.12.0"></a>
# [0.12.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.11.0...v0.12.0) (2018-12-06)


### Features

* do not delete old registry index for speed ([96ff811](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/96ff811))



<a name="0.11.0"></a>
# [0.11.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.10.0...v0.11.0) (2018-10-26)


### Features

* add logo ([1cf85b9](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/1cf85b9))
* allow using non-disposable ipfs nodes ([101e630](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/101e630))


### Performance Improvements

* fall back to npm if ipfs is too slow ([f96a751](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/f96a751))



<a name="0.10.0"></a>
# [0.10.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.9.0...v0.10.0) (2018-10-04)


### Features

* add s3 storage option ([6a522c8](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/6a522c8))



<a name="0.9.0"></a>
# [0.9.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.8.0...v0.9.0) (2018-10-04)


### Bug Fixes

* friendlier error pages ([747320d](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/747320d))


### Features

* use ipfs to fetch files if available ([5e135fe](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/5e135fe))



<a name="0.8.0"></a>
# [0.8.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.7.4...v0.8.0) (2018-09-21)


### Features

* adds retry when npm 404s, fixes [#61](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/61) ([d884991](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/d884991))



<a name="0.7.4"></a>
## [0.7.4](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.7.3...v0.7.4) (2018-09-20)



<a name="0.7.3"></a>
## [0.7.3](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.7.2...v0.7.3) (2018-09-20)


### Bug Fixes

* increase max body size ([d478378](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/d478378))



<a name="0.7.2"></a>
## [0.7.2](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.7.1...v0.7.2) (2018-09-20)


### Bug Fixes

* fixes [#59](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/59) and [#60](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/60) with better error detection and messages ([93ad289](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/93ad289))



<a name="0.7.1"></a>
## [0.7.1](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.7.0...v0.7.1) (2018-09-20)



<a name="0.7.0"></a>
# [0.7.0](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.6.4...v0.7.0) (2018-09-20)


### Features

* forward non-get requests to the registry, fixes [#58](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/58) ([b493e4c](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/b493e4c))



<a name="0.6.4"></a>
## [0.6.4](https://github.com/ipfs-shipyard/npm-on-ipfs/compare/v0.5.1...v0.6.4) (2018-09-20)


### Bug Fixes

* do not use path.join because windows ([8b51156](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/8b51156))
* handle scoped modules, fixes [#57](https://github.com/ipfs-shipyard/npm-on-ipfs/issues/57) ([bd08ddd](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/bd08ddd))
* linting errors ([ca7fd83](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/ca7fd83))
* update arg name ([b707cc0](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/b707cc0))
* use same blobstore instance as it will create an ipfs node ([afe5c58](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/afe5c58))


### Features

* allow configuring external host address separately from internal ([3405f0a](https://github.com/ipfs-shipyard/npm-on-ipfs/commit/3405f0a))

src/cli/bin.js
#! /usr/bin/env node

'use strict'

require('dnscache')({ enable: true })

const pkg = require('../../package')
const path = require('path')
const os = require('os')
const rc = require('rc')

process.title = pkg.name

const yargs = require('yargs')
  .config(rc(pkg.name, null, {}))

yargs.command('$0', 'Installs your js dependencies using IPFS', (yargs) => { // eslint-disable-line no-unused-expressions
  yargs
    .option('package-manager', {
      describe: 'Which package manager to use - eg. npm or yarn',
      default: process.argv[1].indexOf('yarn') === -1 ? 'npm' : 'yarn'
    })

    .option('ipfs-registry', {
      describe: 'Where to download any packages that haven\'t made it into the registry index yet from',
      default: 'https://registry.js.ipfs.io'
    })
    .option('registry-upload-size-limit', {
      describe: 'How large a file upload to allow when proxying for the registry',
      default: '1024MB'
    })
    .option('registry-update-interval', {
      describe: 'Only request the manifest for a given module every so many ms',
      default: 60000
    })
    .option('registry-connect-timeout', {
      describe: 'How long to wait while dialling the mirror before timing out',
      default: 5000
    })
    .option('registry-read-timeout', {
      describe: 'How long to wait for individual packages before timing out',
      default: 5000
    })

    .option('ipfs-mfs-prefix', {
      describe: 'Which mfs prefix to use',
      default: '/npm-registry'
    })
    .option('ipfs-node', {
      describe: '"proc" to start an in-process IPFS node, "disposable" to start an in-process disposable node, "go" or "js" to spawn an IPFS node as a separate process or a multiaddr that resolves to a running node',
      default: 'proc'
    })
    .option('ipfs-repo', {
      describe: 'If --ipfs-node is set to "proc", this is the path that contains the IPFS repo to use',
      default: path.join(os.homedir(), '.jsipfs')
    })
    .option('ipfs-flush', {
      describe: 'Whether to flush the MFS cache',
      default: true
    })

    .option('clone-pin', {
      describe: 'Whether to pin cloned modules',
      default: false
    })

    .option('request-max-sockets', {
      describe: 'How many concurrent http requests to make while cloning the repo',
      default: 10
    })
    .option('request-retries', {
      describe: 'How many times to retry when downloading manifests and tarballs from the registry',
      default: 5
    })
    .option('request-retry-delay', {
      describe: 'How long in ms to wait between retries',
      default: 1000
    })
    .option('request-timeout', {
      describe: 'How long in ms we should wait when requesting files',
      default: 30000
    })

    .option('npm-registry', {
      describe: 'A fallback to use if the IPFS npm registry is unavailable',
      default: 'https://registry.npmjs.com'
    })
}, require('../core'))
  .argv
